### PR TITLE
tests: Add fuzzing and unit tests for ToCompactStr

### PR DESCRIPTION
**Pull request by ParkMyCar**
Like the title says, adds fuzz coverage and more unit tests for testing the `ToCompactStr` trait

### CODE REVIEW

File: compact_str/src/tests.rs [hunk 1]
```diff
@@ -370,6 +370,141 @@ fn test_plus_operator() {
...
+#[should_panic]
+fn test_f64_scientific() {
+    let val: f64 = 1.3594035585e-314;
+
+    let compact = val.to_compact_str();
+    let word = val.to_string();
+
+    assert_eq!(compact, word);
+}
+
```
Comment (NobodyXu) [inline compact_str/src/tests.rs:507]: Just my two cents:

Maybe we should also have `i128` and `u128` tests here?

File: compact_str/src/tests.rs [hunk 2]
```diff
@@ -370,6 +370,141 @@ fn test_plus_operator() {
     assert_eq!(String::from("a") + CompactStr::from("b"), "ab");
 }
```
Comment (NobodyXu) [inline compact_str/src/tests.rs:372]: So we basically unit test the integer and float specialisation of `ToCompactStr` here.

Note that we also have specialisation for `NonZero*` newtype wrappers of the integer, `char`, `String` and `CompactStr`.

File: fuzz/src/lib.rs [hunk 3]
```diff
@@ -34,6 +37,44 @@ pub enum Creation<'a> {
...
     FromBoxStr(Box<str>),
+    /// Create from a number using [`ToCompactStr`]
+    NumToCompactStr(NumType),
+    /// Create from a boolean using [`ToCompactStr`]
+    BoolToCompactStr(bool),
+    /// Create from a char using [`ToCompactStr`]
+    CharToCompactStr(char),
+    /// Create from a String using [`ToCompactStr`]
+    StringToComapctStr(String),
+}
```
Comment (NobodyXu) [inline fuzz/src/lib.rs:48]: Note that we also have specialisation for `CompactStr` in `ToCompactStr`.

File: fuzz/src/lib.rs [hunk 4]
```diff
@@ -180,6 +221,64 @@ impl Creation<'_> {
...
+                    NumType::I128(val) => (val.to_compact_str(), val.to_string()),
+                    // NumType::F32(val) => (val.to_compact_str(), val.to_string()),
+                    // NumType::F64(val) => (val.to_compact_str(), val.to_string()),
+                };
+
+                assert_eq!(compact, word);
+
+                // Numbers should always be inlined, except 64 or 128 bit integers, they might be
+                // heap allocated
+                match num_type {
```
Comment (NobodyXu) [inline fuzz/src/lib.rs:245]: I think it is possible to do this:

```suggestion
                match num_type {
                    NumType::U128(_) | NumType::I128(_) => {
                        assert_properly_allocated(&compact, &word);
                    }
                    #[cfg(target_pointer_width = "32")]
                    NumType::U64(_) | NumType::I64(_) => {
                        assert_properly_allocated(&compact, &word);
                    }
                    _ => assert!(!compact.is_heap_allocated()),
                };
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

**Comment by NobodyXu** [comment]
There are really a ton of changes of binary files for the fuzzing.

File: compact_str/src/tests.rs (hunk 1 above)
Comment (ParkMyCar) [reply compact_str/src/tests.rs:613]: Good idea, added!

File: compact_str/src/tests.rs (hunk 2 above)
Comment (ParkMyCar) [reply compact_str/src/tests.rs:372]: Good call out for the `NonZero*` types, I added those to the fuzzing

File: fuzz/src/lib.rs (hunk 4 above)
Comment (ParkMyCar) [reply fuzz/src/lib.rs]: Changed this to a different general case

File: fuzz/src/lib.rs (hunk 3 above)
Comment (ParkMyCar) [reply fuzz/src/lib.rs:60]: Good callout!

**Comment by ParkMyCar** [comment]
> There are really a ton of changes of binary files for the fuzzing.

Yeah those are the corpus/seed files, there are other strategies for them, e.g. uploading them somewhere like S3, but for now it's just easy to keep them in the repo

**Comment by NobodyXu** [comment]
> > There are really a ton of changes of binary files for the fuzzing.
> 
> Yeah those are the corpus/seed files, there are other strategies for them, e.g. uploading them somewhere like S3, but for now it's just easy to keep them in the repo

If they are auto generated, is it possible for it to be generated in the github action then cached?

**Pull request merged**